<div style="text-align:left;font-size:2em"><span style="font-weight:bolder;font-size:1.25em">SP2273 | Learning Portfolio</span><br><br><span style="font-weight:bold;color:darkred">Loops (Good)</span></div>

## 1 Interrupting the flow

In [3]:
for i in range(10):
    if i>6: break   #forces the loop out of the code
    print(i)

0
1
2
3
4
5
6


In [6]:
for i in range(10):
    if i == 8: continue    #continue works to reject when i == 8, brings it back to the loop and reiterates throughout until it finishes
    print(i)

0
1
2
3
4
5
6
7
9


In [7]:
for i in range(10):
    if i%2 == 0: continue
    print(i)

1
3
5
7
9


In [9]:
number = 0

while True:
    print(number)
    number += 1
    if number >4: break

0
1
2
3
4


## 2 List comprehension!

### 2.1 Basic syntax

In [11]:
[number for number in range(10)]


[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

In [15]:
[number<3 for number in range(10)]  # internal masking from the list

[True, True, True, False, False, False, False, False, False, False]

### 2.2 List comprehension with conditions

In [13]:
[number**2 for number in range(10) if number%2 ==0]


[0, 4, 16, 36, 64]

In [22]:
[number**2 for number in range(10) if number != 4]

[0, 1, 4, 9, 25, 36, 49, 64, 81]

## 3 Other useful stuff

### 3.1 for with unpacking

In [38]:
x, y, _, _, z, _ = [1, 2, 3,4,6,9]

print(f' x = {x}, y = {y}, z = {z}')

py_superhero_info=[['Natasha Romanoff', 'Black Widow'],
                     ['Tony Stark', 'Iron Man'],
                     ['Stephen Strange', 'Doctor Strange']]

for real_name, super_name in py_superhero_info:         #the real_name, super_name refers to the specific positioning within each list of py_superhero info
    print(f"{real_name} is Marvel's {super_name}!")

 x = 1, y = 2, z = 6
Natasha Romanoff is Marvel's Black Widow!
Tony Stark is Marvel's Iron Man!
Stephen Strange is Marvel's Doctor Strange!


### 3.2 for with a zip()

In [32]:
super_names = ["Black Widow", "Iron Man", "Doctor Strange"]
real_names = ["Natasha Romanoff", "Tony Stark", "Stephen Strange"]

for real_name, super_name in zip(real_names,super_names):  #combines 2 lists together in the position of [real_names, super_names]
    print(f"{real_name} is Marvel's {super_name}!")

Natasha Romanoff is Marvel's Black Widow!
Tony Stark is Marvel's Iron Man!
Stephen Strange is Marvel's Doctor Strange!


### 3.3 for with dictionaries

In [ ]:
superhero_info={"Natasha Romanoff": "Black Widow",
                "Tony Stark": "Iron Man",
                "Stephen Strange": "Doctor Strange"}

for key, value in superhero_info.items():
    print(f"{key} is Marvel's {value}!")


## Exercise 1 :  Make me an odd list

In [55]:
%%timeit

list = []
for i in range(10):
    if i%2 == 0: continue
    x = i**2
    list.append(x)
#print(list)





1.27 µs ± 14.2 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


In [54]:
%%timeit
x = 1

5.12 ns ± 0.034 ns per loop (mean ± std. dev. of 7 runs, 100,000,000 loops each)


## Exercise 2 :  Make me another odd list

In [51]:
%%timeit
[number**2 for number in range(10) if number%2 !=0 ]



1.25 µs ± 5.83 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


## Exercise 3 :  Time me!

1.27 µs ± 14.2 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each) in the for loop

1.25 µs ± 5.83 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each) in the list comprehension



